runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -3523,6 +3523,327 @@ J9::ARM64::TreeEvaluator::ArrayCHKEvaluator(TR::Node *node, TR::CodeGenerator *c
    return VMarrayCheckEvaluator(node, cg);
    }
 
+void
+J9::ARM64::TreeEvaluator::genWrtbarForArrayCopy(TR::Node *node, TR::Register *srcObjReg, TR::Register *dstObjReg, TR::CodeGenerator *cg)
+   {
+   TR::Compilation *comp = cg->comp();
+   bool ageCheckIsNeeded;
+   bool cardMarkIsNeeded;
+   auto gcMode = TR::Compiler->om.writeBarrierType();
+
+   ageCheckIsNeeded = (gcMode == gc_modron_wrtbar_oldcheck ||
+                       gcMode == gc_modron_wrtbar_cardmark_and_oldcheck ||
+                       gcMode == gc_modron_wrtbar_always);
+   cardMarkIsNeeded = (gcMode == gc_modron_wrtbar_cardmark ||
+                       gcMode == gc_modron_wrtbar_cardmark_incremental);
+
+   if (!ageCheckIsNeeded && !cardMarkIsNeeded)
+      return;
+
+   if (ageCheckIsNeeded)
+      {
+      TR::Register *tmp1Reg = NULL;
+      TR::Register *tmp2Reg = NULL;
+      TR::RegisterDependencyConditions *deps;
+      TR::Instruction *gcPoint;
+      TR::LabelSymbol *doneLabel;
+
+      if (gcMode != gc_modron_wrtbar_always)
+         {
+         tmp1Reg = cg->allocateRegister();
+         tmp2Reg = cg->allocateRegister();
+         deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(3, 3, cg->trMemory());
+         TR::addDependency(deps, tmp1Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+         TR::addDependency(deps, tmp2Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+         }
+      else
+         {
+         deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(1, 1, cg->trMemory());
+         }
+
+      TR::addDependency(deps, dstObjReg, TR::RealRegister::x0, TR_GPR, cg);
+
+      TR::SymbolReference *wbRef = comp->getSymRefTab()->findOrCreateWriteBarrierBatchStoreSymbolRef(comp->getMethodSymbol());
+
+      if (gcMode != gc_modron_wrtbar_always)
+         {
+         doneLabel = generateLabelSymbol(cg);
+
+         TR::Register *metaReg = cg->getMethodMetaDataRegister();
+
+         // tmp1Reg = heapBaseForBarrierRange0
+         // tmp2Reg = dstObjReg - heapBaseForBarrierRange0
+         generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmx, node, tmp1Reg,
+                                    new (cg->trHeapMemory()) TR::MemoryReference(metaReg, offsetof(J9VMThread, heapBaseForBarrierRange0), cg));
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::subx, node, tmp2Reg, dstObjReg, tmp1Reg);
+
+         // if (tmp2Reg >= heapSizeForBarrierRage0), object not in the tenured area
+         generateCompareInstruction(cg, node, tmp2Reg, tmp1Reg, true);
```

Comment:
tmp1Reg has `heapBaseForBarrierRange0` at this point. I think you need to load `heapSizeForBarrierRange0` before comparison.

Reply:
It seems I mixed up `heapBaseForBarrierRange0` and `heapSizeForBarrierRange0` when I wrote the code.